### PR TITLE
Haichen/route networking2

### PR DESCRIPTION
**Pull request by Haichen-Wang**
## Overview

Justin, Lesley, Aastha, and I finished route networking. 

## Changes Made

**Made route networking work**
 - `getRouteOptions` forms queries when passed in parameters ` start`, `end`, `time`, `arriveBy`, `destinationName`
 - `getRouteOptions` then gets data from the backend and parses the data, ultimately returns a `RouteOptions` object

**Simplified networking code**
- Instead of having many different adapters for each sub-data type, now we only have a `RouteAdapter` and an adapter for parsing a date string into a `Date` object.

## Test Coverage

We ran the code with parameters simulating a real query and got correct output displayed in Logcat.

## Related PRs or Issues (delete if not applicable)
This PR is an update to my old PR https://github.com/cuappdev/ithaca-transit-android-v2/pull/21

### CODE REVIEW

**Review by aasthashah999 — COMMENTED** [review]
I think everything looks good! I think we should restructure some things such as how Coordinate is handled in JsonLocation. I am not sure we need to make additional fields for lat and long. In addition, classes such as Stop and Location have lats and longs. Can we possible make these classes inherit from Coordinate?

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.example.ithaca_transit_android_v2.models
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+//Location. - Represents either a bus stop or a google place.
+@JsonClass(generateAdapter = true)
```
Comment (JiangoJ) [inline app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt]: Change this as this is different from a location

File: app/src/main/java/com/example/ithaca_transit_android_v2/MainActivity.kt [hunk 2]
```diff
@@ -15,15 +15,33 @@ class MainActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        val end = Coordinate(42.444971674516864, -76.48098092526197)
+//        val uid = "E4A0256E-5865-4E9F-8A5A-33747CAC7EBF"
+        val time = 1574292741.0
+        val destinationName = "Bill & Melinda Gates Hall"
+        val start = Coordinate(42.44717985041025, -76.48551732274225)
+        val arriveBy = false
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/MainActivity.kt]: please delete or add a comment saying that this is for testing only

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Direction.kt [hunk 3]
```diff
@@ -1,12 +1,21 @@
...
-    val busStops: List<Location>,
-    val busNumber: Int
+    @Json(name = "startLocation")
+    val startCoords: Coordinate,
+    @Json(name = "endLocation")
+    val endCoords: Coordinate,
+    @Json(name = "stops")
+    val busStops: List<Stop>,
+    @Json(name = "routeNumber")
+    val busNumber: Int?
```
Comment (JiangoJ) [inline app/src/main/java/com/example/ithaca_transit_android_v2/models/Direction.kt:20]: Is there a way that we could prevent getting a null using an adapter?

File: app/src/main/java/com/example/ithaca_transit_android_v2/MainActivity.kt [hunk 4]
```diff
@@ -15,15 +15,33 @@ class MainActivity : AppCompatActivity() {
...
+
+            val maxLogSize = 1000
+            val stringLength = deferred.toString()?.length
+            if (stringLength != null) {
+                for (i in 0..stringLength / maxLogSize) {
+                    val start = i * maxLogSize
+                    var end = (i + 1) * maxLogSize
+                    end = if (end > deferred.toString().length) deferred.toString().length else end
+                    Log.v("returnBody", deferred.toString().substring(start, end))
+                }
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/MainActivity.kt]: please delete or turn this into a helper function

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkUtils.kt [hunk 5]
```diff
@@ -57,4 +59,43 @@ class NetworkUtils {
...
+        json.put("destinationName", destName)
+
+        val requestBody = json.toString().toRequestBody(mediaType)
+        val request: Request = Request.Builder()
+            .url(url + "route")
+            .post(requestBody)
+            .build()
+
+        val body = client.newCall(request).execute().body?.string()
+        Log.d("testing", body)
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkUtils.kt]: delete all log statements

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt [hunk 6]
```diff
@@ -44,4 +48,66 @@ class LocationAdapter {
         }
         return DataLocation(final)
     }
+}
+
+
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: delete extra newlines

File: app/src/main/java/com/example/ithaca_transit_android_v2/MainActivity.kt [hunk 7]
```diff
@@ -15,15 +15,33 @@ class MainActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        val end = Coordinate(42.444971674516864, -76.48098092526197)
+//        val uid = "E4A0256E-5865-4E9F-8A5A-33747CAC7EBF"
+        val time = 1574292741.0
```
Comment (JiangoJ) [inline app/src/main/java/com/example/ithaca_transit_android_v2/MainActivity.kt]: I think we may need to add the UID back in later on?

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt [hunk 8]
```diff
@@ -44,4 +48,66 @@ class LocationAdapter {
...
+        val endCoords: Coordinate,
+        @Json(name ="arrivalTime")
+        val arrival: Date,
+        @Json(name ="departureTime")
+        val depart: Date
+    )
+
+    @FromJson
+    private fun fromJson(json: JsonRoute): Route {
+        fun computeBoardInMin(firstBusDirection: Direction): Int {
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: this method should be moved into the `Route` class

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/DirectionType.kt [hunk 9]
```diff
@@ -1,9 +1,13 @@
 package com.example.ithaca_transit_android_v2.models
 
+import com.squareup.moshi.Json
+
 /*
 Represents the type of route that is displayed on the map
  */
-enum class RouteType {
+enum class DirectionType {
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/models/DirectionType.kt:8]: is this tested? i'm not tried to use `@Json` this way before...

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt [hunk 10]
```diff
@@ -44,4 +48,66 @@ class LocationAdapter {
...
+    private val dateFormat = SimpleDateFormat(SERVER_FORMAT, Locale.getDefault())
+
+    @FromJson
+    fun fromJson(date: String): Date {
+        return dateFormat.parse(date)
+    }
+
+    @ToJson
+    fun toJson(writer: JsonWriter, value: Date?) {
+        if (value != null) {
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: `value?.let { writer.value(value.toString()) }`

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt [hunk 11]
```diff
@@ -44,4 +48,66 @@ class LocationAdapter {
...
+            writer.value(value.toString())
+        }
+    }
+
+    companion object {
+        const val SERVER_FORMAT = ("yyyy-MM-dd'T'HH:mm:ss'Z'") // define your server format here
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: delete the comment

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt [hunk 12]
```diff
@@ -44,4 +48,66 @@ class LocationAdapter {
...
+    }
+
+    @ToJson
+    fun toJson(writer: JsonWriter, value: Date?) {
+        if (value != null) {
+            writer.value(value.toString())
+        }
+    }
+
+    companion object {
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: jw, what's the difference between making a companion object and declaring this object above, like with `dateFormat`

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt [hunk 13]
```diff
@@ -44,4 +48,66 @@ class LocationAdapter {
...
+        return Route(json.directions, json.startCoords, json.endCoords, json.arrival, json.depart,
+            boardInMins)
+    }
+
+    @ToJson
+    private fun toJson(route: Route): JsonRoute {
+        return JsonRoute(emptyList(), Coordinate(0.0,0.0),
+            Coordinate(0.0,0.0), Calendar.getInstance().time,
+            Calendar.getInstance().time)
+    }
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: we shouldn't be using default values; might as well omit this function or leave a `TODO`

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt [hunk 14]
```diff
@@ -44,4 +48,66 @@ class LocationAdapter {
...
+        }
+
+        var firstBus = 0
+        var boardInMins = 0
+        if(json.directions[0].type != DirectionType.BUS){
+            firstBus = 1
+        }
+        if(json.directions.size != 1){
+            boardInMins = computeBoardInMin(json.directions[firstBus])
+        }
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: i feel like we can shorten these two values with a ternary or something, ie 
`var firstBus = if json.direction[0].type == DirectionType.BUS 1 else 0`

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package com.example.ithaca_transit_android_v2.models
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+//Location. - Represents either a bus stop or a google place.
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt]: delete this comment

**Review by ningning621 — CHANGES_REQUESTED** [review]
some initial thoughts, but overall good work!

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+@JsonClass(generateAdapter = true)
+// TODO: deprecate this class 
+data class Stop(
+    val stopID: String,
+    val lat: Double,
+    val long: Double
```
Comment (ningning621) [inline app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt:8]: use `Coordinate`

**Comment by ningning621** [comment]
> I think everything looks good! I think we should restructure some things such as how Coordinate is handled in JsonLocation. I am not sure we need to make additional fields for lat and long. In addition, classes such as Stop and Location have lats and longs. Can we possible make these classes inherit from Coordinate?

i don't think inheriting a class is the right move since `Stop` and `Location` are not subclasses of `Coordinate`.

File: app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt (hunk 12 above)
Comment (Haichen-Wang) [reply app/src/main/java/com/example/ithaca_transit_android_v2/NetworkingAdapters.kt]: We realized making it a companion object was unnecessary and have moved it out.

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt (hunk 16 above)
Comment (Haichen-Wang) [reply app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt:8]: Backend has `stop` in our current format. Their `lat` and `long` are in the same level with `stopID`. Maybe we should leave them as lat and long?

Comment (aasthashah999) [reply app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt:8]: For Stop, the lat and long aren't nested, so replacing lat and long with Coordinate will not work. We tried and got errors. If we really wanted to use the Coordinate, we will have to use it the same way as location by making a Adapter for stops. I am not sure if using Coordinate is worth refactoring the code?

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt (hunk 1 above)
Comment (aasthashah999) [reply app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt]: Im not sure I understand what you mean here

**Review by aasthashah999 — COMMENTED** [review]
I fixed the changes I did not have comments on and resolved them.

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt (hunk 16 above)
Comment (ningning621) [reply app/src/main/java/com/example/ithaca_transit_android_v2/models/Stop.kt:8]: right, it has the same issue as `Location`. it's fine to keep it as is

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/DirectionType.kt (hunk 9 above)
Comment (aasthashah999) [reply app/src/main/java/com/example/ithaca_transit_android_v2/models/DirectionType.kt:8]: When the body was printed in logcat, the correct enums were being identified

**Review by aasthashah999 — APPROVED** [review]
Everything looks good!

File: app/src/main/java/com/example/ithaca_transit_android_v2/models/Direction.kt (hunk 3 above)
Comment (Haichen-Wang) [reply app/src/main/java/com/example/ithaca_transit_android_v2/models/Direction.kt:20]: I think we can leave it as nullable for now. When displaying the bus number, we'll need to check if busNumber is null and don't display it in that case (the only occasion when bus number is null should be a walking direction).

**Review by ningning621 — APPROVED** [review]
lgtm, let's 🚀 it!

**Pull request merged**